### ISSUE TITLE
Merge results and merging v1 with v2 group

### ISSUE THREAD

**Issue by Lensi** [bepaald/signalbackup-tools#26]
Hey I just successfully merged two backups of versions 64 and 89.

These were the errors I could find searching for error, warn and note (anything else I should search for?):

```
SQL error during sqlite3_prepare_v2(): no such table: mention
SQL error during sqlite3_prepare_v2(): no such table: remapped_recipients
SQL error during sqlite3_prepare_v2(): no such table: remapped_threads
  NOTE: Dropping column 'groups.avatar' from source : Column not present in target database
  NOTE: Dropping column 'mms.reveal_start_time' from source : Column not present in target database
  NOTE: Skipping table 'megaphone', as it is not present in target database. Data may be missing.
  Dealing with table 'part'... 121/946 entries...Warning: attachment data not found (rowid: 121, uniqueid: 1600536479253)
```

Let me know if you need more info or want me to run it again later.

Also, I have a group that changed id after being upgraded to a v2 group. Before I look into renaming it in the old backup, maybe you have some advice? It's such a small thing, but I will give it a try unless you know it's more complex than I hope :)

Thank you for a wonderful tool!

P.S. I couldn't find your email address.

**Comment by Lensi** [comment]
Can some of the errors be fixed by reimporting the old backup into a new Signal version before merging?

**Comment by bepaald** [comment]
Thanks for your report. I *think* none of the errors, notes and warnings would be a problem, but the sql errors do surprise me, I wonder where they come from. If possible, I would like to have a few lines of context so I can maybe track down where in the program these errors occur.

By the way, which backup was the target and which was the source? Usually it's smart to use the newer database as target and the older one as source.

> Also, I have a group that changed id after being upgraded to a v2 group. Before I look into renaming it in the old backup, maybe you have some advice? It's such a small thing, but I will give it a try unless you know it's more complex than I hope :)

Yes, I had thought about this case. I'm not sure what you mean by 'renaming' but the merge does not use the group name to find the target group, instead it uses the (internal) id of the group. I can think of two ways to merge the V1 and V2 groups.

 - One: change the id of the V1 group in the old database before merging:
```
./signalbackup-tools --fast --runsqlquery 'UPDATE groups SET group_id = "__signal_group__v2__!....." WHERE group_id == "__textsecure_group__!...."' --runsqlquery 'UPDATE recipient SET group_id = "__signal_group__v2__!....." WHERE group_id == "__textsecure_group__!..."' [INPUT] [PASSWORD] -o [OUTPUT]
```
 - Two: after merging, move all messages to the v2 group and delete the v1 group:
```
./signalbackup-tools --fast --runsqlquery 'UPDATE sms SET thread_id = NEWTHREADID WHERE thread_id = OLDTHREADID'  --runsqlquery 'UPDATE mms SET thread_id = NEWTHREADID WHERE thread_id = OLDTHREADID' --runsqlquery 'DELETE FROM thread WHERE _id = OLDTHREADID' --runsqlquery 'DELETE FROM recipient WHERE group_id == "__textsecure_group__!....."' --runsqlquery 'DELETE FROM groups WHERE group_id == "__textsecure_group__!....."' [INPUT] [PASSWORD] -o [OUTPUT]
```

**Both untested**. I think I'd go for the first option, but I'm not sure why. For both command the necessary thread_id's and group_id's can be determined from `signalbackup-tools --listthreads [INPUT] [PASSWORD]`. For case two, only run this on the merged output, and use those values. For case 1, run this on both backups to get the new (signal_group_v2) and old (textsecure_group) ids from the databases.

>Thank you for a wonderful tool!

No problem!

>P.S. I couldn't find your email address.

Yes. I should probably put it in the README somewhere. Thanks. (It's basjetimmer [ at ] yahoo-dot-com by the way.


> Can some of the errors be fixed by reimporting the old backup into a new Signal version before merging?

Yes probably, the signal database changes rapidly and those two versions are quite a bit apart. However, I don't expect it to have caused problems. Have you imported the backup yet? How did it look?

**Comment by bepaald** [comment]
I think I managed to find the place where the errors appeared. They shouldn't have changed the outcome of the merge, but I have changed the code to check for tables/columns before trying to alter them. Thanks!

**Comment by Lensi** [comment]
Cool, I can confirm your update removed the three SQL errors. The rest remain as expected, with the notes appearing for every thread.

I used the older database as source.

I have imported the merged database into Signal and it looks great, however I quickly reverted to give it another go along with merging the groups as mentioned (I am not aware of a way to casually test import on secondary phone without reregistering with Signal).

I will get back to you after I attempt the group chat merge.
